File: src/main/java/org/cloudfoundry/credhub/util/TimedRetry.java
package org.cloudfoundry.credhub.util;

import java.util.function.Supplier;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

@Component
public class TimedRetry {

  public static final int ONE_SECOND = 1000;
  private final CurrentTimeProvider currentTimeProvider;

  @Autowired
  public TimedRetry(final CurrentTimeProvider currentTimeProvider) {
    super();
    this.currentTimeProvider = currentTimeProvider;
  }

  public boolean retryEverySecondUntil(final long durationInSeconds, final Supplier<Boolean> untilTrue) {
    final long startTime = currentTimeProvider.currentTimeMillis();
    long currentTime;
    final long endTime = startTime + ONE_SECOND * durationInSeconds;

    do {
      if (untilTrue.get()) {
        return true;
      }
      try {
        currentTimeProvider.sleep(ONE_SECOND);
      } catch (final InterruptedException e) {
        // do nothing until we want to use InterruptedExceptions to
        // cause graceful shutdowns
      }

      currentTime = currentTimeProvider.currentTimeMillis();
    } while (currentTime < endTime);

    return false;
  }
}

File: components/time-support/src/main/kotlin/org/cloudfoundry/credhub/util/TimedRetry.kt
package org.cloudfoundry.credhub.util

import java.util.function.Supplier

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class TimedRetry @Autowired
constructor(private val currentTimeProvider: CurrentTimeProvider) {

    fun retryEverySecondUntil(durationInSeconds: Long, untilTrue: Supplier<Boolean>): Boolean {
        val startTime = currentTimeProvider.currentTimeMillis()
        var currentTime: Long
        val endTime = startTime + ONE_SECOND * durationInSeconds

        do {
            if (untilTrue.get()) {
                return true
            }
            try {
                currentTimeProvider.sleep(ONE_SECOND.toLong())
            } catch (e: InterruptedException) {
                // do nothing until we want to use InterruptedExceptions to
                // cause graceful shutdowns
            }

            currentTime = currentTimeProvider.currentTimeMillis()
        } while (currentTime < endTime)

        return false
    }

    companion object {

        val ONE_SECOND = 1000
    }
}
